stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -327,9 +327,13 @@ function _sparsesimilar(S::SparseMatrixCSC, ::Type{TvNew}, ::Type{TiNew}) where
     newrowval = copyto!(similar(S.rowval, TiNew), S.rowval)
     return SparseMatrixCSC(S.m, S.n, newcolptr, newrowval, similar(S.nzval, TvNew))
 end
-# parent methods for similar that preserves only storage space (for when new and old dims differ)
-_sparsesimilar(S::SparseMatrixCSC, ::Type{TvNew}, ::Type{TiNew}, dims::Dims{2}) where {TvNew,TiNew} =
-    SparseMatrixCSC(dims..., fill(one(TiNew), last(dims)+1), similar(S.rowval, TiNew), similar(S.nzval, TvNew))
+# parent methods for similar that preserves only storage space allocation (for when new and old dims differ)
+function _sparsesimilar(S::SparseMatrixCSC, ::Type{TvNew}, ::Type{TiNew}, dims::Dims{2}) where {TvNew,TiNew}
+    S1=spzeros(TvNew, TiNew, dims...)
+    sizehint!(S1.rowval, min(length(S.rowval), length(S1)))
+    sizehint!(S1.nzval, min(length(S.nzval), length(S1)))
```

Comment:
Might be nicer to define a `sizehint!` method for SparseMatrix (and SparseVector), and then do `sizehint(S1, min(length(S.nzval), length(S1)))`.  This would make it easier for the user to create sparse matrices with custom size hints.

Also, are similar changes needed for sparsevector?

Reply:
`similar(::SparseVector, ::Integer) :: SparseVector` didn't allocate, but you are right in that `similar(SparseVector, ::Dims{2}) :: SparseMatrixCSC` has the same behaviour. I did both changes.